File: lib/media.py
import subprocess

from lib.helpers import run


def ffprobe_duration(path):
    result = subprocess.check_output([
        "ffprobe",
        "-v", "error",
        "-show_entries", "format=duration",
        "-of", "default=noprint_wrappers=1:nokey=1",
        path,
    ])
    return float(result.decode().strip())


def merge_remove_segments(remove_segments):
    if not remove_segments:
        return []

    segments = sorted(
        [(segment["start"], segment["end"]) for segment in remove_segments],
        key=lambda item: item[0],
    )

    merged = [segments[0]]

    for start, end in segments[1:]:
        last_start, last_end = merged[-1]

        if start <= last_end:
            merged[-1] = (last_start, max(last_end, end))
        else:
            merged.append((start, end))

    return merged


def build_keep_segments(duration, remove_segments):
    merged_remove = merge_remove_segments(remove_segments)

    if not merged_remove:
        return [(0.0, duration)]

    keep = []
    current = 0.0

    for start, end in merged_remove:
        start = max(0.0, min(start, duration))
        end = max(0.0, min(end, duration))

        if start > current:
            keep.append((current, start))

        current = max(current, end)

    if current < duration:
        keep.append((current, duration))

    return keep


def build_hybrid_subsegments(keep_segment, remove_segments, boundary_window):
    keep_start, keep_end = keep_segment
    keep_duration = keep_end - keep_start
    if keep_duration <= 0:
        return []

    left_adjacent = any(abs(segment["end"] - keep_start) < 0.001 for segment in remove_segments)
    right_adjacent = any(abs(segment["start"] - keep_end) < 0.001 for segment in remove_segments)

    if not left_adjacent and not right_adjacent:
        return [{
            "start": keep_start,
            "end": keep_end,
            "cut_mode": "copy",
        }]

    if keep_duration <= boundary_window * 2:
        return [{
            "start": keep_start,
            "end": keep_end,
            "cut_mode": "precise",
        }]

    segments = []
    current_start = keep_start

    if left_adjacent:
        left_end = min(keep_end, keep_start + boundary_window)
        segments.append({
            "start": keep_start,
            "end": left_end,
            "cut_mode": "precise",
        })
        current_start = left_end

    middle_end = keep_end
    if right_adjacent:
        middle_end = max(current_start, keep_end - boundary_window)

    if middle_end > current_start:
        segments.append({
            "start": current_start,
            "end": middle_end,
            "cut_mode": "copy",
        })

    if right_adjacent:
        right_start = max(keep_start, keep_end - boundary_window)
        if right_start < keep_end:
            if segments and segments[-1]["cut_mode"] == "copy" and segments[-1]["end"] > right_start:
                segments[-1]["end"] = right_start
            segments.append({
                "start": right_start,
                "end": keep_end,
                "cut_mode": "precise",
            })

    return [segment for segment in segments if segment["end"] - segment["start"] > 0]


def cap_subsegment_durations(subsegments, max_duration_seconds):
    if max_duration_seconds is None:
        return list(subsegments)

    max_duration_seconds = float(max_duration_seconds)
    if max_duration_seconds <= 0:
        raise RuntimeError("segment_max_render_seconds must be greater than 0")

    capped = []
    for subsegment in subsegments:
        start = float(subsegment["start"])
        end = float(subsegment["end"])
        cut_mode = subsegment["cut_mode"]

        if end <= start:
            continue

        current_start = start
        while current_start < end:
            current_end = min(end, current_start + max_duration_seconds)
            capped.append({
                "start": current_start,
                "end": current_end,
                "cut_mode": cut_mode,
            })
            current_start = current_end

    return capped


def get_keyframes(video_path):
    """Get sorted list of keyframe PTS timestamps from a video file.

    Uses ffprobe to detect all keyframes (I-frames). Empty list on error.
    """
    try:
        cmd = [
            "ffprobe", "-v", "error",
            "-select_streams", "v:0",
            "-show_entries", "packet=pts_time",
            "-of", "csv=p=0",
            "-skip_frame", "nokey",
            str(video_path),
        ]
        result = subprocess.check_output(cmd, encoding="utf-8", errors="replace").strip()
        if not result:
            return []
        times = []
        for line in result.split("\n"):
            line = line.strip()
            if line:
                try:
                    times.append(float(line))
                except ValueError:
                    pass
        return times
    except (subprocess.CalledProcessError, FileNotFoundError, OSError):
        return []


def align_subsegments_to_keyframes(subsegments, keyframes):
    """Align copy-mode subsegment boundaries to keyframes to prevent overlap.

    With -c copy, -ss before -i seeks to the nearest keyframe *before* the
    requested time, causing consecutive copy subsegments to overlap by up to
    the keyframe interval.

    This only touches subsegments with cut_mode="copy". Precise subsegments
    (re-encoded) pass through unchanged.

    Internal copy-to-copy boundaries are aligned so adjacent subsegments meet
    at the *same* keyframe. Edges that touch a precise segment are preserved:
    - first copy keeps its original start (interface with left precise)
    - last copy keeps its original end   (interface with right precise)

    This works for both pure-copy mode (all subsegments are "copy") and hybrid
    mode (mix of "copy" and "precise").
    """
    if not keyframes or not subsegments:
        return subsegments

    keyframes = sorted(set(keyframes))

    def kf_before(time):
        result = None
        for kf in keyframes:
            if kf <= time + 0.001:
                result = kf
            else:
                break
        return result

    aligned = []

    for i, sub in enumerate(subsegments):
        if sub["cut_mode"] != "copy":
            aligned.append(sub)
            continue

        start = sub["start"]
        end = sub["end"]

        prev_is_copy = aligned and aligned[-1]["cut_mode"] == "copy"
        if prev_is_copy:
            kf_start = kf_before(start)
            if kf_start is None:
                kf_start = start
            prev_end = aligned[-1]["end"]
            kf_start = max(kf_start, prev_end)
        else:
            kf_start = start

        next_is_copy = i + 1 < len(subsegments) and subsegments[i + 1]["cut_mode"] == "copy"
        kf_end = end
        if next_is_copy:
            aligned_end = kf_before(end)
            if aligned_end is not None:
                kf_end = aligned_end

        if kf_end > kf_start:
            aligned.append({
                "start": kf_start,
                "end": kf_end,
                "cut_mode": sub["cut_mode"],
            })

    return aligned


def render_segment_copy(ep_file, segment_output, start, end):
    cmd = [
        "ffmpeg",
        "-y",
        "-ss", f"{start:.3f}",
        "-to", f"{end:.3f}",
        "-i", ep_file,
        "-map", "0:v:0",
        "-map", "0:a:0?",
        "-map", "0:s?",
        "-c", "copy",
        "-c:s", "copy",
        segment_output,
    ]

    run(cmd)


def render_segment_precise(ep_file, segment_output, start, end, segment_encoding=None):
    duration = max(0.0, end - start)
    if duration <= 0:
        raise RuntimeError(f"Invalid precise segment duration for {ep_file}: {start} -> {end}")

    encoding = segment_encoding or {}
    video_codec = encoding.get("video_codec", "libx264")
    preset = encoding.get("preset", "medium")
    cq = str(encoding.get("cq", 18 if "nvenc" in video_codec else 15))
    audio_codec = encoding.get("audio_codec", "aac")
    audio_bitrate = encoding.get("audio_bitrate", "192k")
    pixel_format = encoding.get("pixel_format")

    cmd = [
        "ffmpeg",
        "-y",
        "-ss", f"{start:.3f}",
        "-to", f"{end:.3f}",
        "-i", ep_file,
        "-map", "0:v:0",
        "-map", "0:a:0?",
        "-map", "0:s?",
        "-c:v", video_codec,
    ]

    if "nvenc" in video_codec:
        cmd += ["-preset", preset, "-cq", cq]
    elif video_codec in ["libx264", "libx265"]:
        cmd += ["-preset", preset, "-crf", cq]

    if pixel_format:
        cmd += ["-pix_fmt", pixel_format]

    cmd += [
        "-c:a", audio_codec,
        "-b:a", audio_bitrate,
        "-c:s", "copy",
        segment_output,
    ]

    run(cmd)


def render_segment(ep_file, segment_output, start, end, segment_encoding=None):
    duration = max(0.0, end - start)
    if duration <= 0:
        raise RuntimeError(f"Invalid segment duration for {ep_file}: {start} -> {end}")

    cut_mode = (segment_encoding or {}).get("cut_mode", "hybrid")
    if cut_mode == "precise":
        render_segment_precise(ep_file, segment_output, start, end, segment_encoding=segment_encoding)
        return

    render_segment_copy(ep_file, segment_output, start, end)


def render_concat(concat_file, concat_output):
    fast_path = [
        "ffmpeg",
        "-y",
        "-f", "concat",
        "-safe", "0",
        "-i", concat_file,
        "-map", "0",
        "-c", "copy",
        concat_output,
    ]

    try:
        run(fast_path)
        return
    except subprocess.CalledProcessError:
        pass

    safe_path = [
        "ffmpeg",
        "-y",
        "-fflags", "+genpts",
        "-f", "concat",
        "-safe", "0",
        "-i", concat_file,
        "-map", "0:v?",
        "-map", "0:a:0?",
        "-map", "0:s?",
        "-c:v", "libx264",
        "-preset", "ultrafast",
        "-crf", "18",
        "-c:a", "aac",
        "-c:s", "copy",
        concat_output,
    ]
    run(safe_path)


def render_final(concat_output, watermark_path, output_video, encoding):
    video_codec = encoding.get("video_codec", "h264_nvenc")
    preset = encoding.get("preset", "fast")
    cq = str(encoding.get("cq", 23))
    audio_codec = encoding.get("audio_codec", "aac")

    overlay_filter = (
        "[0:v]format=yuv420p[base];"
        "[1:v]scale=160:-1,format=rgba[wm];"
        "[base][wm]overlay=W-w-20:20,format=yuv420p[v]"
    )

    cmd = [
        "ffmpeg",
        "-y",
        "-i", concat_output,
        "-i", watermark_path,
        "-filter_complex", overlay_filter,
        "-map", "[v]",
        "-map", "0:a:0?",
        "-map", "0:s?",
        "-c:v", video_codec,
    ]

    if "nvenc" in video_codec:
        cmd += ["-preset", preset, "-cq", cq]
    elif video_codec in ["libx264", "libx265"]:
        cmd += ["-preset", preset, "-crf", cq]

    cmd += [
        "-c:a", audio_codec,
        "-c:s", "copy",
        output_video,
    ]

    run(cmd)

File: src/core/media.ts
import { exec, run } from "../shared/helpers";
import type { AudioStream, RemoveSegment, KeepSegment, Subsegment } from "../shared/types";

// ============================================================================
// Audio stream detection
// ============================================================================

export function detectAudioStreams(videoPath: string): AudioStream[] {
  try {
    const result = Bun.spawnSync([
      "ffprobe", "-v", "error",
      "-select_streams", "a",
      "-show_entries", "stream=index:stream_tags=language,title,handler_name:stream_disposition=default",
      "-of", "json",
      videoPath,
    ], { stdout: "pipe", stderr: "pipe" });

    if (result.exitCode !== 0) return [];

    const data = JSON.parse(new TextDecoder().decode(result.stdout));
    const streams: AudioStream[] = [];

    for (let audioIndex = 0; audioIndex < (data.streams ?? []).length; audioIndex++) {
      const stream = data.streams[audioIndex];
      const tags = stream.tags ?? {};
      const disposition = stream.disposition ?? {};
      streams.push({
        audio_index: audioIndex,
        stream_index: stream.index ?? 0,
        language: tags.language ?? null,
        title: tags.title ?? null,
        handler_name: tags.handler_name ?? null,
        is_default: Boolean(disposition.default ?? 0),
      });
    }
    return streams;
  } catch {
    return [];
  }
}

function buildLanguageVariants(preferred: string): Set<string> {
  const normalized = preferred.trim().toLowerCase();
  const aliasMap: Record<string, Set<string>> = {
    rus: new Set(["rus", "ru", "russian", "russkiy", "рус", "русский", "дубляж", "озвучка"]),
    ru: new Set(["rus", "ru", "russian", "russkiy", "рус", "русский", "дубляж", "озвучка"]),
    russian: new Set(["rus", "ru", "russian", "russkiy", "рус", "русский", "дубляж", "озвучка"]),
    jpn: new Set(["jpn", "jp", "japanese", "nihongo", "япон", "японский"]),
    jp: new Set(["jpn", "jp", "japanese", "nihongo", "япон", "японский"]),
    japanese: new Set(["jpn", "jp", "japanese", "nihongo", "япон", "японский"]),
  };
  return aliasMap[normalized] ?? (normalized ? new Set([normalized]) : new Set());
}

export function getPreferredAudioStream(videoPath: string, preferredLanguage = "rus"): number {
  const streams = detectAudioStreams(videoPath);
  if (streams.length === 0) return 0;

  const preferredVariants = buildLanguageVariants(preferredLanguage);
  let bestMatch: { audioIndex: number; score: number } | null = null;

  for (const stream of streams) {
    let score = 0;
    const lang = stream.language?.toLowerCase();
    const title = stream.title?.toLowerCase();
    const handler = stream.handler_name?.toLowerCase();

    if (lang && preferredVariants.has(lang)) score = 3;
    else if (title && preferredVariants.has(title)) score = 2;
    else if (handler && preferredVariants.has(handler)) score = 1;

    if (stream.is_default) score += 1;

    if (score > 0 && (!bestMatch || score > bestMatch.score)) {
      bestMatch = { audioIndex: stream.audio_index, score };
    }
  }

  if (bestMatch) return bestMatch.audioIndex;

  // Return default stream if any
  for (const stream of streams) {
    if (stream.is_default) return stream.audio_index;
  }

  return 0;
}

// ============================================================================
// ffprobe
// ============================================================================

export function ffprobeDuration(videoPath: string): number {
  const result = Bun.spawnSync([
    "ffprobe", "-v", "error",
    "-show_entries", "format=duration",
    "-of", "default=noprint_wrappers=1:nokey=1",
    videoPath,
  ], { stdout: "pipe", stderr: "pipe" });

  if (result.exitCode !== 0) {
    throw new Error(`ffprobe failed on ${videoPath}`);
  }

  return parseFloat(new TextDecoder().decode(result.stdout).trim());
}

// ============================================================================
// Keyframes
// ============================================================================

export function getKeyframes(videoPath: string): number[] {
  const result = Bun.spawnSync([
    "ffprobe", "-v", "error",
    "-select_streams", "v:0",
    "-show_entries", "packet=pts_time,flags",
    "-of", "csv=p=0",
    videoPath,
  ], { stdout: "pipe", stderr: "pipe" });

  if (result.exitCode !== 0) return [];

  const lines = new TextDecoder().decode(result.stdout).trim().split("\n").filter(Boolean);
  const keyframes: number[] = [0];

  for (const line of lines) {
    const parts = line.split(",");
    if (parts.length >= 2) {
      const pts = parseFloat(parts[0]!);
      const flags = parts[1]!;
      if (flags.includes("K") && !isNaN(pts) && pts > 0) {
        keyframes.push(pts);
      }
    }
  }

  return keyframes;
}

// ============================================================================
// Segment manipulation
// ============================================================================

export function snapRemoveSegmentsToKeyframes(
  removeSegments: RemoveSegment[],
  keyframes: number[],
): RemoveSegment[] {
  if (keyframes.length === 0) return removeSegments;

  return removeSegments.map(seg => {
    let snapStart = seg.start;
    let snapEnd = seg.end;

    for (const kf of keyframes) {
      if (kf <= seg.start) snapStart = kf;
    }
    for (let i = keyframes.length - 1; i >= 0; i--) {
      if (keyframes[i]! >= seg.end) snapEnd = keyframes[i]!;
    }

    return { ...seg, start: snapStart, end: snapEnd };
  });
}

export function buildKeepSegments(
  duration: number,
  removeSegments: RemoveSegment[],
): KeepSegment[] {
  const sorted = [...removeSegments].sort((a, b) => a.start - b.start);
  const keep: KeepSegment[] = [];
  let current = 0;

  for (const seg of sorted) {
    if (seg.start > current) {
      keep.push({ start: current, end: seg.start });
    }
    current = Math.max(current, seg.end);
  }

  if (current < duration) {
    keep.push({ start: current, end: duration });
  }

  return keep;
}

export function buildHybridSubsegments(
  interval: KeepSegment,
  removeSegments: RemoveSegment[],
  boundaryWindow: number,
): Subsegment[] {
  const subsegments: Subsegment[] = [];
  let pos = interval.start;

  for (const seg of removeSegments.filter(s => s.start >= interval.start && s.end <= interval.end).sort((a, b) => a.start - b.start)) {
    // Before removed segment
    const beforeEnd = seg.start;
    if (beforeEnd > pos) {
      const beforeStart = Math.max(pos, beforeEnd - boundaryWindow);
      if (beforeStart < beforeEnd) {
        if (beforeStart > pos) {
          subsegments.push({ start: pos, end: beforeStart, cut_mode: "copy" });
        }
        subsegments.push({ start: beforeStart, end: beforeEnd, cut_mode: "precise" });
      } else {
        subsegments.push({ start: pos, end: beforeEnd, cut_mode: "copy" });
      }
    }
    pos = seg.end;

    // After removed segment
    const afterEnd = Math.min(interval.end, pos + boundaryWindow);
    if (afterEnd > pos) {
      subsegments.push({ start: pos, end: afterEnd, cut_mode: "precise" });
      pos = afterEnd;
    }
  }

  // Remaining
  if (pos < interval.end) {
    subsegments.push({ start: pos, end: interval.end, cut_mode: "copy" });
  }

  return mergeAdjacentCopySubsegments(subsegments);
}

function mergeAdjacentCopySubsegments(subsegments: Subsegment[]): Subsegment[] {
  if (subsegments.length <= 1) return subsegments;
  const merged: Subsegment[] = [];
  let current = subsegments[0]!;

  for (let i = 1; i < subsegments.length; i++) {
    const next = subsegments[i]!;
    if (current.cut_mode === "copy" && next.cut_mode === "copy") {
      current = { start: current.start, end: next.end, cut_mode: "copy" };
    } else {
      merged.push(current);
      current = next;
    }
  }
  merged.push(current);
  return merged;
}

export function capSubsegmentDurations(
  subsegments: Subsegment[],
  maxSeconds: number,
): Subsegment[] {
  const result: Subsegment[] = [];
  for (const seg of subsegments) {
    const duration = seg.end - seg.start;
    if (duration <= maxSeconds) {
      result.push(seg);
    } else {
      // Split into chunks
      for (let pos = seg.start; pos < seg.end; pos += maxSeconds) {
        const end = Math.min(seg.end, pos + maxSeconds);
        result.push({ start: pos, end, cut_mode: seg.cut_mode });
      }
    }
  }
  return result;
}

// ============================================================================
// Rendering
// ============================================================================

export function renderSegment(
  inputPath: string,
  outputPath: string,
  start: number,
  end: number,
  options: {
    segmentEncoding?: Record<string, unknown>;
    audioStreamIndex?: number;
  } = {},
): void {
  const encoding = options.segmentEncoding ?? {};
  const cutMode = (encoding.cut_mode as string) ?? "precise";
  const videoCodec = (encoding.video_codec as string) ?? "libx264";
  const preset = (encoding.preset as string) ?? "medium";
  const cq = String(encoding.cq ?? 18);
  const audioCodec = (encoding.audio_codec as string) ?? "aac";
  const audioBitrate = (encoding.audio_bitrate as string) ?? "192k";
  const pixelFormat = (encoding.pixel_format as string) || "yuv420p";
  const ai = options.audioStreamIndex ?? 0;

  const args: string[] = [
    "ffmpeg", "-y", "-v", "error",
    "-ss", start.toFixed(3),
    "-to", (end - start).toFixed(3),
    "-i", inputPath,
    "-map", "0:v:0",
    "-map", `0:a:${ai}`,
  ];

  if (cutMode === "copy") {
    args.push("-c", "copy");
  } else {
    args.push(
      "-c:v", videoCodec,
      "-preset", preset,
      "-cq", cq,
      "-pix_fmt", pixelFormat,
      "-c:a", audioCodec,
      "-b:a", audioBitrate,
    );
  }

  args.push(outputPath);
  run(args);
}

export function renderConcat(concatFilePath: string, outputPath: string): void {
  run([
    "ffmpeg", "-y", "-v", "error",
    "-f", "concat",
    "-safe", "0",
    "-i", concatFilePath,
    "-c", "copy",
    outputPath,
  ]);
}

export function renderFinal(
  concatOutputPath: string,
  watermarkPath: string,
  outputVideoPath: string,
  options: {
    encoding?: Record<string, unknown>;
    audioStreamIndex?: number;
  } = {},
): void {
  const encoding = options.encoding ?? {};
  const videoCodec = (encoding.video_codec as string) ?? "h264_nvenc";
  const preset = (encoding.preset as string) ?? "fast";
  const cq = String(encoding.cq ?? 23);
  const audioCodec = (encoding.audio_codec as string) ?? "aac";
  const ai = options.audioStreamIndex ?? 0;

  const args: string[] = [
    "ffmpeg", "-y", "-v", "error",
    "-i", concatOutputPath,
    "-i", watermarkPath,
    "-map", "0:v:0",
    "-map", `0:a:${ai}`,
  ];

  // Apply watermark overlay with NVENC
  args.push(
    "-filter_complex", "[1:v]format=rgba,colorchannelmixer=aa=0.5[wm];[0:v][wm]overlay=W-w-10:H-h-10:format=auto,format=yuv420p",
    "-c:v", videoCodec,
    "-preset", preset,
    "-cq", cq,
    "-c:a", audioCodec,
    outputVideoPath,
  );

  run(args);
}
